official/recommendation/data_pipeline.py

```diff
@@ -0,0 +1,727 @@
+    mask has been computed.
+
+    Args:
+      users: An array of users in a batch. (should be identical along axis 1)
+      positive_items: An array (batch_size x 1) of positive item indices.
+      negative_items: An array of negative item indices.
+      users_per_batch: How many users should be in the batch. This is passed
+        as an argument so that ncf_test.py can use this method.
+
+    Returns:
+      User, item, and duplicate_mask arrays.
+    """
+    items = np.concatenate([positive_items, negative_items], axis=1)
+
+    # We pad the users and items here so that the duplicate mask calculation
+    # will include the padding. The metric function relies on every element
+    # except the positive being marked as duplicate to mask out padded points.
+    if users.shape[0] < users_per_batch:
+      pad_rows = users_per_batch - users.shape[0]
+      padding = np.zeros(shape=(pad_rows, users.shape[1]), dtype=np.int32)
+      users = np.concatenate([users, padding.astype(users.dtype)], axis=0)
+      items = np.concatenate([items, padding.astype(items.dtype)], axis=0)
+
+    duplicate_mask = stat_utils.mask_duplicates(items, axis=1).astype(np.bool)
+
+    items[:, (0, -1)] = items[:, (-1, 0)]
+    duplicate_mask[:, (0, -1)] = duplicate_mask[:, (-1, 0)]
+
+    assert users.shape == items.shape == duplicate_mask.shape
+    return users, items, duplicate_mask
+
+  def _get_eval_batch(self, i):
+    """Construct a single batch of evaluation data.
+
+    Args:
+      i: The index of the batch.
+    """
+    low_index = i * self._eval_users_per_batch
+    high_index = (i + 1) * self._eval_users_per_batch
+    users = np.repeat(self._eval_pos_users[low_index:high_index, np.newaxis],
+                      1 + rconst.NUM_EVAL_NEGATIVES, axis=1)
+    positive_items = self._eval_pos_items[low_index:high_index, np.newaxis]
+    negative_items = (self.lookup_negative_items(negative_users=users[:, :-1])
+                      .reshape(-1, rconst.NUM_EVAL_NEGATIVES))
+
+    users, items, duplicate_mask = self._assemble_eval_batch(
+        users, positive_items, negative_items, self._eval_users_per_batch)
+
+    self._eval_dataset.put(i, {
+        movielens.USER_COLUMN: users.flatten(),
+        movielens.ITEM_COLUMN: items.flatten(),
+        rconst.DUPLICATE_MASK: duplicate_mask.flatten(),
+    })
+
+  def _construct_eval_epoch(self):
+    """Loop to construct data for evaluation."""
+    if self._stop_loop:
+      return
+
+    start_time = timeit.default_timer()
+
+    self._eval_dataset.start_construction()
+    map_args = [i for i in range(self.eval_batches_per_epoch)]
+    with popen_helper.get_threadpool(6) as pool:
+      pool.map(self._get_eval_batch, map_args)
+    self._eval_dataset.end_construction()
+
+    tf.logging.info("Eval construction complete. Time: {:.1f} seconds".format(
+        timeit.default_timer() - start_time))
+
+  def make_input_fn(self, is_training):
+    if self._fatal_exception is not None:
```

Comment:
A fatal exception can occur after this check, in which case you won't catch it and the program will likely hang. This is hard to fix, but maybe comment that this is a bit racy.